From cae62dbb51a5812374b015e7782b2d5ad976e8e3 Mon Sep 17 00:00:00 2001
From: George Cherian <george.cherian@ti.com>
Date: Tue, 17 Dec 2013 14:19:31 +0530
Subject: [PATCH 033/188] usb: musb: dsps: Enhance the Babble interrupt
 handling

This commit comes from branch ti-linux-3.12.y:

git://git.ti.com/ti-linux-kernel/ti-linux-kernel.git

During babble interrupt make sure the phy power is switched off
and then turned on. This gives some time for the babbling device
to settle down and also reset the device.

Also reinitialize the controller endpoints and fifo adresses.

Signed-off-by: George Cherian <george.cherian@ti.com>
(cherry picked from commit d27d50b62a3c71899e85c75686c4f5cbef1372e0)
---
 drivers/usb/musb/musb_core.c |    3 +++
 drivers/usb/musb/musb_dsps.c |    6 ++++++
 2 files changed, 9 insertions(+), 0 deletions(-)

diff --git a/drivers/usb/musb/musb_core.c b/drivers/usb/musb/musb_core.c
index f97aeae..ecd1339 100644
--- a/drivers/usb/musb/musb_core.c
+++ b/drivers/usb/musb/musb_core.c
@@ -1604,6 +1604,9 @@ EXPORT_SYMBOL_GPL(musb_interrupt);
 
 void musb_babble_reinit(struct musb *musb)
 {
+	musb_core_init(musb->config->multipoint
+			? MUSB_CONTROLLER_MHDRC
+			: MUSB_CONTROLLER_HDRC, musb);
 	musb_start(musb);
 }
 EXPORT_SYMBOL_GPL(musb_babble_reinit);
diff --git a/drivers/usb/musb/musb_dsps.c b/drivers/usb/musb/musb_dsps.c
index 66a3eb9..f6de92c 100644
--- a/drivers/usb/musb/musb_dsps.c
+++ b/drivers/usb/musb/musb_dsps.c
@@ -332,7 +332,13 @@ static void sw_babble_control(struct musb *musb)
 		/* Reset the controller */
 		dsps_writel(base, wrp->control, (1 << wrp->reset));
 
+		usb_phy_shutdown(musb->xceiv);
+		mdelay(100);
 		musb_platform_set_mode(musb, MUSB_HOST);
+		mdelay(100);
+		usb_phy_init(musb->xceiv);
+		mdelay(100);
+
 		musb_babble_reinit(musb);
 	}
 }
-- 
1.7.5.4

